include/ocpp/v201/evse_manager.hpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2024 Pionix GmbH and Contributors to EVerest
+
+#pragma once
+
+#include <ocpp/common/utils.hpp>
+#include <ocpp/v201/evse.hpp>
+
+namespace ocpp {
+namespace v201 {
+
+class EvseOutOfRangeException : public std::exception {
+public:
+    explicit EvseOutOfRangeException(int32_t id) : msg{"Evse with id " + std::to_string(id) + " does not exist"} {
+    }
+
+    ~EvseOutOfRangeException() noexcept override = default;
+
+    const char* what() const noexcept override {
+        return msg.c_str();
+    }
+
+private:
+    std::string msg;
+};
+
+class EvseManagerInterface {
+private:
+    virtual std::vector<std::unique_ptr<EvseInterface>>& get_vector() = 0;
+    virtual const std::vector<std::unique_ptr<EvseInterface>>& get_vector() const = 0;
+
+public:
+    using EvseIterator = VectorOfUniquePtrIterator<EvseInterface>;
+
+    virtual ~EvseManagerInterface() = default;
+
+    EvseIterator begin();
```

Comment:
I am missing documentation (not only here, but for the whole EvseIterator), why you use this and not just a vector. If someone looks at this later, it is very unclear why this is for (and I do not totally understand it either as well, is it really a problem what you are trying to solve here?).

Reply:
I agree this all needs a bit of documentation so I will add that.

The reason I have done it like this is so that we no longer expose the `unique_ptr` that we use just for storage to the "outside" world.

Initially I was also afraid that if you return the vector others could reset or assign the `unique_ptr`s for example. Returning a `const vector` will take care of that, but you still have to deal with pointers instead of references which I think is not ideal.

Also, now you can use range based for on the `evse_manager` in `charge_point`, otherwise you would have to call `get_vector` or a similar function.